language/move-prover/bytecode-to-boogie/src/bytecode_translator.rs

```diff
@@ -1175,14 +1152,32 @@ impl<'env> ModuleTranslator<'env> {
                     if ty.is_mutual_reference() {
                         mutual_local_borrows.insert(*src as usize);
                         if self.has_update_invariant_on_release(&ty) {
-                            update_invariant_on_release.push(ty)
+                            local_to_before_borrow_idx.insert(*dst, num_mut_refs);
+                            num_mut_refs += 1;
                         }
                     }
                 }
-                BorrowGlobal(dst, ..) => {
+                BorrowGlobal(dst, ..) | BorrowField(dst, ..) => {
                     let ty = self.get_local_type(func_env, *dst);
                     if ty.is_mutual_reference() && self.has_update_invariant_on_release(&ty) {
-                        update_invariant_on_release.push(ty)
+                        local_to_before_borrow_idx.insert(*dst, num_mut_refs);
+                        num_mut_refs += 1;
+                    }
+                }
+                MoveLoc(dst, src) => {
+                    // If src has update invariants to check for then insert the dst to
+                    // the map and point to the same before_borrow index so that we can
+                    // later use dst to find the before_borrow index
+                    let src_idx = *src as usize;
+                    if local_to_before_borrow_idx.contains_key(&src_idx) {
+                        let idx = local_to_before_borrow_idx[&src_idx];
+                        local_to_before_borrow_idx.insert(*dst, idx);
+                    }
+                }
+                StLoc(dst, src) => {
+                    if local_to_before_borrow_idx.contains_key(src) {
+                        let idx = local_to_before_borrow_idx[src];
+                        local_to_before_borrow_idx.insert(*dst as usize, idx);
                     }
                 }
                 _ => {}
```

Comment:
Should we also enforce on `Ret` (for references currently on the stack and in locals)?

Reply:
Right. We are enforcing on WriteRef, FreezeRef and Call right now(see lifetime_analysis.rs). I'm working on adding CopyLoc, Pop and Ret.